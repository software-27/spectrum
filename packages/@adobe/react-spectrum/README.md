# @adobe/react-spectrum

This package allows you to easily import any component published under @react-spectrum. Think of it as a helpful alias,
with each component "pinned" to a specific version.

Components are updated sporadically, so if you require a change from a version of a component that is not yet included
in this package, you are able to install the specific version and then update your import statement to use the
component from the @react-spectrum scope. It's worth noting that this may result in increased file size.

## Naming conventions

Any component published under the @react-spectrum scope will be available at the same path from @adobe/react-spectrum.
For example, if you want to use `Checkbox` from @react-spectrum this is what you'd be importing:

```js
import {Checkbox} from '@react-spectrum/checkbox';
```

To import the same component from @adobe/react-spectrum, it would look like this:

```js
import {Checkbox} from '@adode/react-spectrum/checkbox';
```

This convention will be consistent for all packages to make it easy to switch between @react-spectrum and
@adobe/react-spectrum.

## Component versions

| Component                          | Version        |
| ---------------------------------- | -------------- |
| @react-spectrum/actiongroup        | 3.0.0-alpha.1  |
| @react-spectrum/breadcrumbs        | 3.0.0-alpha.1  |
| @react-spectrum/button             | 3.0.0-rc.2     |
| @react-spectrum/buttongroup        | 3.0.0-alpha.1  |
| @react-spectrum/checkbox           | 3.0.0-rc.2     |
| @react-spectrum/dialog             | 3.0.0-alpha.1  |
| @react-spectrum/divider            | 3.0.0-rc.2     |
| @react-spectrum/form               | 3.0.0-rc.2     |
| @react-spectrum/icon               | 3.0.0-rc.2     |
| @react-spectrum/illustratedmessage | 3.0.0-alpha.1  |
| @react-spectrum/image              | 3.0.0-alpha.1  |
| @react-spectrum/label              | 3.0.0-rc.2     |
| @react-spectrum/layout             | 3.0.0-alpha.1  |
| @react-spectrum/link               | 3.0.0-alpha.1  |
| @react-spectrum/listbox            | 3.0.0-alpha.1  |
| @react-spectrum/menu               | 3.0.0-alpha.1  |
| @react-spectrum/meter              | 3.0.0-rc.2     |
| @react-spectrum/overlays           | 3.0.0-alpha.1  |
| @react-spectrum/picker             | 3.0.0-alpha.1  |
| @react-spectrum/progress           | 3.0.0-rc.2     |
| @react-spectrum/provider           | 3.0.0-rc.2     |
| @react-spectrum/radio              | 3.0.0-rc.2     |
| @react-spectrum/searchfield        | 3.0.0-rc.2     |
| @react-spectrum/statuslight        | 3.0.0-rc.2     |
| @react-spectrum/switch             | 3.0.0-rc.2     |
| @react-spectrum/textfield          | 3.0.0-rc.2     |
| @react-spectrum/theme-default      | 3.0.0-rc.2     |
| @react-spectrum/utils              | 3.0.0-rc.2     |
| @react-spectrum/view               | 3.0.0-alpha.1  |
| @react-spectrum/well               | 3.0.0-rc.2     |
