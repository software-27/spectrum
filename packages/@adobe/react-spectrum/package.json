{
  "name": "@adobe/react-spectrum",
  "version": "3.0.0-rc.0",
  "description": "Spectrum UI components in React",
  "license": "Apache-2.0",
  "private": true,
  "main": "dist/main.js",
  "module": "dist/module.js",
  "types": "dist/types.d.ts",
  "source": "src/index.ts",
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "scripts": {
    "prepublishOnly": "mkdir -p dist; cp src/index.ts dist/types.d.ts; cp src/index.ts dist/module.js; babel --root-mode upward src/index.ts -o dist/main.js"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/adobe-private/react-spectrum-v3"
  },
  "dependencies": {
    "@react-spectrum/actiongroup": "^3.0.0",
    "@react-spectrum/breadcrumbs": "^3.0.0",
    "@react-spectrum/button": "^3.0.0",
    "@react-spectrum/buttongroup": "^3.0.0",
    "@react-spectrum/checkbox": "^3.0.0",
    "@react-spectrum/dialog": "^3.0.0",
    "@react-spectrum/divider": "^3.0.0",
    "@react-spectrum/form": "^3.0.0",
    "@react-spectrum/icon": "^3.0.0",
    "@react-spectrum/illustratedmessage": "^3.0.0",
    "@react-spectrum/image": "^3.0.0",
    "@react-spectrum/label": "^3.0.0",
    "@react-spectrum/layout": "^3.0.0",
    "@react-spectrum/link": "^3.0.0",
    "@react-spectrum/listbox": "^3.0.0",
    "@react-spectrum/menu": "^3.0.0",
    "@react-spectrum/meter": "^3.0.0",
    "@react-spectrum/overlays": "^3.0.0",
    "@react-spectrum/picker": "^3.0.0",
    "@react-spectrum/progress": "^3.0.0",
    "@react-spectrum/provider": "^3.0.0",
    "@react-spectrum/radio": "^3.0.0",
    "@react-spectrum/searchfield": "^3.0.0",
    "@react-spectrum/statuslight": "^3.0.0",
    "@react-spectrum/switch": "^3.0.0",
    "@react-spectrum/text": "^3.0.0",
    "@react-spectrum/textfield": "^3.0.0",
    "@react-spectrum/theme-dark": "^3.0.0",
    "@react-spectrum/theme-default": "^3.0.0",
    "@react-spectrum/view": "^3.0.0",
    "@react-spectrum/well": "^3.0.0"
  }
}
