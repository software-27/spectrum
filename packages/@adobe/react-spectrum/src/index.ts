/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

export * from '@react-spectrum/actiongroup';
export * from '@react-spectrum/breadcrumbs';
export * from '@react-spectrum/button';
export * from '@react-spectrum/buttongroup';
export * from '@react-spectrum/checkbox';
export * from '@react-spectrum/dialog';
export * from '@react-spectrum/divider';
export * from '@react-spectrum/form';
export * from '@react-spectrum/icon';
export * from '@react-spectrum/illustratedmessage';
export * from '@react-spectrum/image';
export * from '@react-spectrum/label';
export * from '@react-spectrum/layout';
export * from '@react-spectrum/link';
export * from '@react-spectrum/listbox';
export * from '@react-spectrum/menu';
export * from '@react-spectrum/meter';
export * from '@react-spectrum/overlays';
export * from '@react-spectrum/picker';
export * from '@react-spectrum/progress';
export * from '@react-spectrum/provider';
export * from '@react-spectrum/radio';
export * from '@react-spectrum/searchfield';
export * from '@react-spectrum/statuslight';
export * from '@react-spectrum/switch';
export * from '@react-spectrum/text';
export * from '@react-spectrum/textfield';
export {theme as darkTheme} from '@react-spectrum/theme-dark';
export {theme as defaultTheme} from '@react-spectrum/theme-default';
export * from '@react-spectrum/view';
export * from '@react-spectrum/well';
