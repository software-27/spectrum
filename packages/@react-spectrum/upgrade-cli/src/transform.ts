import path from 'node:path';
import {run as jscodeshift} from 'jscodeshift/src/Runner.js';

const transformPath = path.join(__dirname, 'codemods', 'codemod.js');

interface Options {
  componentSubset?: string,
  path?: string,
  ignorePattern?: string,
  parser?: string
}

export async function transform(options: Options) {
  let {
    path: filePath = '.',
    ...rest
  } = options;

  return await jscodeshift(transformPath, [filePath], rest);
}
