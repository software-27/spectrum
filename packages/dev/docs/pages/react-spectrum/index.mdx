<!-- Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. -->

import {Layout} from '@react-spectrum/docs';
export default Layout;

import {Highlights, Hero} from '@react-spectrum/docs';
import heroNarrow from 'url:../assets/hero-narrow.png';
import heroNarrow2x from 'url:../assets/hero-narrow@2x.png';
import heroWide from 'url:../assets/hero-wide.png';
import heroWide2x from 'url:../assets/hero-wide@2x.png';

# React Spectrum

A React implementation of Spectrum, Adobeâ€™s design system.

<Hero
  narrow={heroNarrow}
  narrow2x={heroNarrow2x}
  wide={heroWide}
  wide2x={heroWide2x}
  alt="React Spectrum hero image" />

<Highlights items={[
  {
    title: 'Accessible',
    description: 'React Spectrum components are designed with accessibility in mind, including full screen reader and keyboard navigation support.',
  },
  {
    title: 'Adaptive',
    description: 'React Spectrum components are designed to work with mouse, touch, and keyboard interactions. Built with responsive design principles to deliver a great experience no matter the device.',
  },
  {
    title: 'Themeable',
    description: 'React Spectrum supports custom themes to match your brand, including automatic support for dark mode.'
  },
  {
    title: 'Internationalized',
    description: 'React Spectrum components are designed to support over 30 languages, including support for right to left, date and number formatting, and more.'
  }
]} />
