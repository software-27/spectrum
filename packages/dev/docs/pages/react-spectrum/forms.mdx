{/* Copyright 2023 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. */}

import {Layout} from '@react-spectrum/docs';
export default Layout;

---
category: Concepts
keywords: [forms, validation, frameworks, next.js, remix, react server actions]
order: 5.5
---

# Forms

Forms allow users to enter and submit data, and provide them with feedback along the way. React Spectrum includes many components that integrate with HTML forms, with support for custom validation, labels, and help text.

## Labels and help text

Accessible forms start with clear, descriptive labels for each field. All React Spectrum form components support labeling using the `label` prop. In addition, help text associates additional context with a field such as a description or error message. The label and help text are announced by assistive technology such as screen readers when the user focuses the field.

```tsx example
import {TextField} from '@adobe/react-spectrum';

<TextField type="password" label="Password" description="Password must be at least 8 characters." />
```

Most fields should have a visible label. In rare exceptions, the `aria-label` or `aria-labelledby` attribute must be provided instead to identify the element to screen readers.

## Submitting data

How you submit form data depends on your framework, application, and server. By default, HTML forms are submitted by the browser using a full page refresh. You can take control of form submission by calling `preventDefault` during the `onSubmit` event, and make an API call to submit the data however you like. Frameworks like [Next.js](https://nextjs.org/docs/app/building-your-application/data-fetching/forms-and-mutations) and [Remix](https://remix.run/docs/en/main/components/form) also include builtin APIs to manage this for you.

### Uncontrolled forms

The simplest way to get data from a form is using the browser's [FormData](https://developer.mozilla.org/en-US/docs/Web/API/FormData) API during the `onSubmit` event. This can be passed directly to [fetch](https://developer.mozilla.org/en-US/docs/Web/API/fetch), or converted into a regular JavaScript object using [Object.fromEntries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries). Each field should have a `name` prop to identify it, and values will be serialized to strings by the browser.

```tsx example
import {Form, TextField, ButtonGroup, Button} from '@adobe/react-spectrum';

function Example() {
  let [submitted, setSubmitted] = React.useState(null);

  /*- begin highlight -*/
  let onSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    // Prevent default browser page refresh.
    e.preventDefault();

    // Get form data as an object.
    let data = Object.fromEntries(new FormData(e.currentTarget));

    // Submit to your backend API...
    setSubmitted(data);
  };
  /*- end highlight -*/

  return (
    <Form onSubmit={onSubmit} maxWidth="size-3000">
      <TextField name="name" label="Name" />
      <ButtonGroup>
        <Button type="submit" variant="primary">Submit</Button>
        <Button type="reset" variant="secondary">Reset</Button>
      </ButtonGroup>
      {submitted && <div>You submitted: <code>{JSON.stringify(submitted)}</code></div>}
    </Form>
  );
}
```

### Controlled forms

By default, all React Spectrum components are uncontrolled, which means that the state is stored internally on your behalf. If you need access to the value in realtime, as the user is editing, you can make it controlled. You'll need to manage the state using React's [useState](https://react.dev/reference/react/useState) hook, and pass the current value and a change handler into each form component.

```tsx example
import {Form, TextField, ButtonGroup, Button} from '@adobe/react-spectrum';

function Example() {
  /*- begin highlight -*/
  let [name, setName] = React.useState('');
  /*- end highlight -*/

  let onSubmit = (e) => {
    e.preventDefault();

    // Submit data to your backend API...
    alert(name);
  };

  return (
    <Form onSubmit={onSubmit} maxWidth="size-3000">
      {/*- begin highlight -*/}
      <TextField label="Name" value={name} onChange={setName} />
      {/*- end highlight -*/}
      <div>You entered: {name}</div>
      <ButtonGroup>
        <Button type="submit" variant="primary">Submit</Button>
        <Button type="reset" variant="secondary">Reset</Button>
      </ButtonGroup>
    </Form>
  );
}
```

## Validation

Form validation is important to ensure user input is in an expected format and meets business requirements. Well-designed form validation assists the user with specific, helpful error messages without confusing and frustrating them with unnecessary errors on partial input. React Spectrum supports native HTML constraint validation with customizable UI, custom validation functions, realtime validation, and integration with server-side validation errors.

### Built-in validation

All React Spectrum form components integrate with native HTML [constraint validation](https://developer.mozilla.org/en-US/docs/Web/HTML/Constraint_validation). This allows you to define constraints on each field such as required, minimum and maximum values, text formats such as email addresses, and even custom regular expression patterns. These constraints are checked by the browser when the user commits changes to the value (e.g. on blur) or submits the form.

To enable native validation, set the `validationBehavior="native"` prop on the [Form](Form.html) component. This example shows a required email field, which is validated by the browser and displayed with Spectrum help text.

```tsx example
<Form validationBehavior="native" maxWidth="size-3000">
  {/*- begin highlight -*/}
  <TextField label="Email" name="email" type="email" isRequired />
  {/*- end highlight -*/}
  <ButtonGroup>
    <Button type="submit" variant="primary">Submit</Button>
    <Button type="reset" variant="secondary">Reset</Button>
  </ButtonGroup>
</Form>
```

Supported constraints include:

* `isRequired` indicates that a field must have a value before the form can be submitted.
* `minValue` and `maxValue` specify the minimum and maximum value in a date picker or number field.
* `minLength` and `maxLength` specify the minimum and length of text input.
* `pattern` provides a custom [regular expression](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions) that a text input must conform to.
* `type="email"` and `type="url"` provide builtin validation for email addresses and URLs.

See each component's documentation for more details on the supported validation props.

### Customizing error messages

By default, React Spectrum displays the error message provided by the browser, which is localized in the user's preferred language. You can customize these messages by providing a function to the `errorMessage` prop. This receives a list of error strings along with a [ValidityState](https://developer.mozilla.org/en-US/docs/Web/API/ValidityState) object describing why the field is invalid, and should return an error message to display.

```tsx example
<Form validationBehavior="native" maxWidth="size-3000">
  <TextField
    label="Name"
    name="name"
    isRequired
    /*- begin highlight -*/
    errorMessage={({validationDetails}) => (
      validationDetails.valueMissing ? 'Please enter a name.' : ''
    )}
    /*- end highlight -*/
  />
  <ButtonGroup>
    <Button type="submit" variant="primary">Submit</Button>
    <Button type="reset" variant="secondary">Reset</Button>
  </ButtonGroup>
</Form>
```

**Note**: The default error messages are localized by the browser using the browser/operating system language setting. React Spectrum's [Provider](Provider.html) has no effect on validation errors.

### Custom validation

In addition to the built-in constraints, custom validation is supported by providing a function to the `validate` prop. This function receives the current field value, and can return a string or array of strings representing one or more error messages. These are displayed to the user after the value is committed (e.g. on blur) to avoid distracting them on each keystroke.

```tsx example
<Form validationBehavior="native" maxWidth="size-3000">
  <TextField
    label="Username"
    /*- begin highlight -*/
    validate={value => value === 'admin' ? 'Nice try!' : null}
    /*- end highlight -*/
  />
  <ButtonGroup>
    <Button type="submit" variant="primary">Submit</Button>
    <Button type="reset" variant="secondary">Reset</Button>
  </ButtonGroup>
</Form>
```

### Realtime validation

Usually, validation errors should be displayed to the user after the value is committed (e.g. on blur), or when the form is submitted. This avoids confusing the user with irrelevant errors while they are still entering a value.

In some cases, validating in realtime can be desireable, such as when meeting password requirements. This can be accomplished by making the field value [controlled](#controlled-forms), and setting the `isInvalid` and `errorMessage` props accordingly.

```tsx example
function Example() {
  let [password, setPassword] = React.useState('');
  let error;
  if (password.length < 8) {
    error = 'Password must be 8 characters or more.';
  } else if ((password.match(/[A-Z]/g) ?? []).length < 2) {
    error = 'Password must include at least 2 upper case letters';
  } else if ((password.match(/[^a-z]/ig) ?? []).length < 2) {
    error = 'Password must include at least 2 symbols.';
  }

  return (
    <TextField
      label="Password"
      /*- begin highlight -*/
      isInvalid={!!error}
      errorMessage={error}
      /*- end highlight -*/
      value={password}
      onChange={setPassword} />
  );
}
```

By default, the `isInvalid` prop does not block forms from being submitted. To block form submission, use the `validationBehavior="native"` prop on the [Form](Form.html) component.

### Server validation

Client side validation is useful to give the user immediate feedback, but only one half of the validation story. Data should also be validated on the backend for security and reliability, and your business logic may include rules which cannot be validated on the frontend.

React Spectrum supports displaying server validation errors by passing the `validationErrors` prop to the [Form](Form.html) component. This should be set to an object that maps each field's `name` prop to a string or array of strings representing one or more errors. These are displayed to the user as soon as the `validationErrors` prop is set, and cleared after the user modifies each field's value.

```tsx example
function Example() {
  /*- begin highlight -*/
  let [errors, setErrors] = React.useState({});
  /*- end highlight -*/
  let onSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    let data = Object.fromEntries(new FormData(e.currentTarget));
    let result = await callServer(data)
    /*- begin highlight -*/
    setErrors(result.errors);
    /*- end highlight -*/
  };

  return (
    <Form
      validationBehavior="native"
      maxWidth="size-3000"
      onSubmit={onSubmit}
      /*- begin highlight -*/
      validationErrors={errors}
      /*- end highlight -*/
    >
      <TextField label="Username" name="username" isRequired />
      <TextField label="Password" name="password" type="password" isRequired />
      <ButtonGroup>
        <Button type="submit" variant="primary">Submit</Button>
        <Button type="reset" variant="secondary">Reset</Button>
      </ButtonGroup>
    </Form>
  );
}

// Fake server used in this example.
function callServer(data) {
  return {
    errors: {
      username: 'Sorry, this username is taken.'
    }
  };
}
```

#### Schema validation

React Spectrum is compatible with errors returned from schema validation libraries like [Zod](https://zod.dev/), which are often used for server-side form validation. Use the [flatten](https://zod.dev/ERROR_HANDLING?id=flattening-errors) method to get a list of errors for each field and return this as part of your HTTP response.

```tsx
// In your server...
import {z} from 'zod';

const schema = z.object({
  name: z.string().min(1),
  age: z.coerce.number().positive()
});

function handleRequest(formData: FormData) {
  let result = schema.safeParse(Object.fromEntries(formData));
  if (!result.success) {
    return {
      /*- begin highlight -*/
      errors: result.error.flatten().fieldErrors
      /*- end highlight -*/
    };
  }

  // Do stuff...

  return {
    errors: {}
  };
}
```

Note that error message localization is also best done on the server rather than on the client to avoid large bundles. You can submit the user's locale as part of the form data if needed, and use something like [zod-i18n](https://github.com/aiji42/zod-i18n) to translate the errors into the correct language.

#### React Server Actions

[Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/forms-and-mutations) are a new React feature currently supported by Next.js. They enable the client to seamlessly call the server without setting up any API routes, and integrate with forms via the `action` prop. The [useFormState](https://react.dev/reference/react-dom/hooks/useFormState) hook can be used to get the value returned by a server action after submitting a form, which may include validation errors.

```tsx
// app/add-form.tsx
'use client';

import {useFormState} from 'react-dom';
import {Form, TextField, ActionButton} from '@adobe/react-spectrum';
import {createTodo} from '@/app/actions';

export function AddForm() {
  /*- begin highlight -*/
  let [{errors}, formAction] = useFormState(createTodo, {errors: {}});
  /*- end highlight -*/

  return (
    /*- begin highlight -*/
    <Form action={formAction} validationErrors={errors}>
    {/*- end highlight -*/}
      <TextField label="Task" name="todo" />
      <ActionButton type="submit">Add</ActionButton>
    </Form>
  );
}
```

```tsx
// app/actions.ts
'use server';

export async function createTodo(prevState: any, formData: FormData) {
  try {
    // Create the todo...
  } catch (err) {
    return {
      errors: {
        /*- begin highlight -*/
        todo: 'Invalid todo.'
        /*- end highlight -*/
      }
    };
  }
}
```

#### Remix

[Remix actions](https://remix.run/docs/en/main/route/action) handle form submissions on the server. The [useSubmit](https://remix.run/docs/en/main/hooks/use-submit) hook can be used to submit data to the server, and the [useActionData](https://remix.run/docs/en/main/hooks/use-action-data) hook can be used to get the value returned by the server, which may include validation errors.

```tsx
// app/routes/signup.tsx
import type {ActionFunctionArgs} from '@remix-run/node';
import {useActionData, useSubmit} from '@remix-run/react';
import {Form, TextField, Button} from '@adobe/react-spectrum';

export default function SignupForm() {
  let submit = useSubmit();
  let onSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    submit(e.currentTarget);
  };

  /*- begin highlight -*/
  let actionData = useActionData<typeof action>();
  /*- end highlight -*/

  return (
    <Form
      method="post"
      /*- begin highlight -*/
      validationErrors={actionData?.errors}
      /*- end highlight -*/
      onSubmit={onSubmit}>
      <TextField label="Username" name="username" isRequired />
      <TextField label="Password" name="password" type="password" isRequired />
      <Button type="submit" variant="cta">Submit</Button>
    </Form>
  );
}

export async function action({request}: ActionFunctionArgs) {
  try {
    // Validate data and perform action...
  } catch (err) {
    return {
      errors: {
        /*- begin highlight -*/
        username: 'Sorry, this username is taken.'
        /*- end highlight -*/
      }
    };
  }
}
```

## Form libraries

In most cases, uncontrolled forms with the builtin validation features are enough. However, if you are building a truly complex form, or integrating React Spectrum components into an existing form, a separate form library such as [React Hook Form](https://react-hook-form.com/) or [Formik](https://formik.org/) may be helpful.

### React Hook Form

[React Hook Form](https://react-hook-form.com/) is a popular form library for React. It is primarily designed to work directly with plain HTML input elements, but supports custom form components like the ones in React Spectrum as well.

Since React Spectrum manages the state for components internally, you can use the [Controller](https://react-hook-form.com/docs/usecontroller/controller) component from React Hook Form to integrate React Spectrum components. Pass the props for the `field` render prop through to the React Spectrum component you're using, and use the `fieldState` to get validation errors to display.

```tsx
import {useForm, Controller} from 'react-hook-form'
import {Form, TextField, Button} from '@adobe/react-spectrum';

function App() {
  let {handleSubmit, control} = useForm({
    defaultValues: {
      name: '',
    },
  });
  let onSubmit = (data) => {
    // Call your API here...
  };

  return (
    <Form onSubmit={handleSubmit(onSubmit)}>
      <Controller
        control={control}
        name="name"
        rules={{ required: 'Name is required.' }}
        render={({
          field: { name, value, onChange, onBlur, ref },
          fieldState: { invalid, error },
        }) => (
          <TextField
            label="Name"
            name={name}
            value={value}
            onChange={onChange}
            onBlur={onBlur}
            ref={ref}
            isRequired
            isInvalid={invalid}
            errorMessage={error?.message}
          />
        )}
      />
      <Button type="submit" variant="cta">Submit</Button>
    </Form>
  );
}
```
