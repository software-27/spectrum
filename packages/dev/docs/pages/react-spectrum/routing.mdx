{/* Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. */}

import {Layout} from '@react-spectrum/docs';
export default Layout;

import {Keyboard} from '@react-spectrum/text';

---
category: Concepts
keywords: [router, links, frameworks, next.js, remix, react-router]
order: 5
---

# Client Side Routing

Many React Spectrum components support rendering as HTML links. This page discusses how to set up your app to integrate React Spectrum links with your framework or client side router.

## Introduction

React Spectrum components such as [Menu](Menu.html), [Tabs](Tabs.html), [TableView](TableView.html), and many others support rendering elements as links that perform navigation when the user interacts with them. Each component that supports link behavior accepts the `href` prop, which causes the component to render an `<a>` element. Other link DOM props such as `target` and `download` are also supported.

By default, links perform native browser navigation when they are interacted with. However, many apps and frameworks use client side routers to avoid a full page reload when navigating between pages. The [Provider](Provider.html) component can configure all React Spectrum components within it to navigate using the client side router you provide. Set this up once in the root of your app, and any React Spectrum component with the `href` prop will automatically navigate using your router.

Note that external links to different origins will not trigger client side routing, and will use native browser navigation. Additionally, if the link has a [target](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#target) other than `"_self"`, uses the [download](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#download) attribute, or the user presses modifier keys such as <Keyboard>Command</Keyboard> or <Keyboard>Alt</Keyboard> to change the default behavior, browser native navigation will occur instead of client side routing.

## Provider setup

The `Provider` component accepts a `router` prop, which can be set to a router object. This should include a `navigate` function received from your router for performing a client side navigation programmatically. The following example shows the general pattern. Framework-specific examples are shown below.

```tsx
import {Provider, defaultTheme} from '@adobe/react-spectrum';

function App() {
  let navigate = useNavigateFromYourRouter();

  return (
    <Provider theme={defaultTheme} router={{navigate}}>
      {/* ... */}
    </Provider>
  );
}
```

### React Router

The [useNavigate](https://reactrouter.com/en/main/hooks/use-navigate) hook from `react-router-dom` returns a `navigate` function you can pass to `Provider`. Ensure that the component that calls `useNavigate` and renders `Provider` is inside the router component (e.g. `BrowserRouter`) so that it has access to React Router's internal context. The React Router `<Routes>` element should also be defined inside React Spectrum's `<Provider>` so that links inside the rendered routes have access to the router.

```tsx
import {BrowserRouter, useNavigate} from 'react-router-dom';
import {Provider, defaultTheme} from '@adobe/react-spectrum';

function App() {
  let navigate = useNavigate();

  return (
    <Provider theme={defaultTheme} router={{navigate}}>
      {/* Your app here... */}
      <Routes>
        <Route path="/" element={<HomePage />} />
        {/* ... */}
      </Routes>
    </Provider>
  );
}

<BrowserRouter>
  <App />
</BrowserRouter>
```

### Next.js

#### App router

The [useRouter](https://nextjs.org/docs/app/api-reference/functions/use-router) hook from `next/navigation` returns a router object that can be used to perform navigation. `Provider` should be rendered from a client component at the root of each page or layout that includes React Spectrum components. You can create a new client component for this, or combine it with other top-level providers as described in the [Next.js docs](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#using-context-providers).

```tsx
// app/provider.tsx
"use client";

import {useRouter} from 'next/navigation';
import {Provider, defaultTheme} from '@adobe/react-spectrum';

export function ClientProviders({children}) {
  let router = useRouter();

  return (
    <Provider theme={defaultTheme} router={{navigate: router.push}}>
      {children}
    </Provider>
  );
}
```

Then, in your page or layout server component, wrap your app in the `ClientProviders` component that you defined.

```tsx
// app/layout.tsx
import {ClientProviders} from './provider';

export default function RootLayout({children}) {
  return (
    <html>
      <body>
        <ClientProviders>{children}</ClientProviders>
      </body>
    </html>
  );
}
```

#### Pages router

The [useRouter](https://nextjs.org/docs/pages/api-reference/functions/use-router) hook from `next/router` returns a router object that can be used to perform navigation. `Provider` should be rendered at the root of each page that includes React Spectrum components, or in `pages/_app.tsx` to add it to all pages.

```tsx
// pages/_app.tsx
import type {AppProps} from 'next/app';
import {useRouter} from 'next/router';
import {Provider, defaultTheme} from '@adobe/react-spectrum';

export default function MyApp({Component, pageProps}: AppProps) {
  let router = useRouter();

  return (
    <Provider theme={defaultTheme} router={{navigate: router.push}}>
      <Component {...pageProps} />
    </Provider>
  );
}
```

### Remix

Remix uses React Router under the hood, so the same [useNavigate](https://reactrouter.com/en/main/hooks/use-navigate) hook described above also works in Remix apps. `Provider` should be rendered at the root of each page that includes React Spectrum components, or in `app/root.tsx` to add it to all pages. See the [Remix docs](https://remix.run/docs/en/main/file-conventions/root) for more details.

```tsx
// app/root.tsx
import {useNavigate, Outlet} from '@remix-run/react';
import {Provider, defaultTheme} from '@adobe/react-spectrum';

export default function App() {
  let navigate = useNavigate();

  return (
    <html lang="en">
      <head>
        {/* ... */}
      </head>
      <body>
        <Provider theme={defaultTheme} router={{navigate}}>
          <Outlet />
        </Provider>
        {/* ... */}
      </body>
    </html>
  );
}
```
