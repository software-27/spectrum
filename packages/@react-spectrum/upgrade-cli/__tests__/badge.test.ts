// @ts-ignore
import {defineSnapshotTest} from 'jscodeshift/dist/testUtils';
import transform from '../src/codemods/codemod';

const test = (name: string, input: string) => {
  defineSnapshotTest(transform, {}, input, name);
};

test('No change', `
import {Badge} from '@adobe/react-spectrum';

<div>
  <Badge variant="positive">Licensed</Badge>
</div>
`);
