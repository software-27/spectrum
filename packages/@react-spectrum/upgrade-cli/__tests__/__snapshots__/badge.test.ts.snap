// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`No change 1`] = `
"import { Badge } from "@react-spectrum/s2";

<div>
  <Badge variant="positive">Licensed</Badge>
</div>"
`;
